commands/peers/addpeer.go

```diff
@@ -31,27 +48,29 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	localNode := false
-	for _, addr := range req.Addresses {
-		local, _ := utils.IsLocalAddress(addr)
-		if local == true {
-			localNode = true
-			break
-		}
-	}
+	// A peer can have multiple addresses. For now, we use only the first
+	// address present in the req.Addresses list.
 
-	if localNode == true {
-		rest.SendHTTPError(w, http.StatusInternalServerError, errors.ErrPeerLocalNode.Error())
+	remotePeerAddress, err := utils.FormRemotePeerAddress(req.Addresses[0])
+	if err != nil {
+		rest.SendHTTPError(w, http.StatusBadRequest, err.Error())
 		return
 	}
 
-	// This remote call will return the remote peer's ID (UUID) and Name.
-	remotePeer, e := ValidateAddPeer(&req)
+	// This remote call will return the remote peer's ID (UUID), name
+	// and etcd peer url.
+	remotePeer, e := ValidateAddPeer(remotePeerAddress, &req)
 	if e != nil {
 		rest.SendHTTPError(w, http.StatusInternalServerError, remotePeer.OpError)
 		return
 	}
 
+	// TODO: Parse addresses considering ports to figure this out.
+	if isPeerInCluster(remotePeer.UUID) {
+		rest.SendHTTPError(w, http.StatusInternalServerError, errors.ErrPeerLocalNode.Error())
```

Comment:
> The error here should be that the new address is already a peer.

 Will change that :+1:  Was an oversight when the local node check was modified.

> is it possible to do a similar check before doing the RPC

Yes. I've already captured this in the TODO comment above the call to `isPeerInCluster`. Will address that later

> Check is the given address is local. If it is return 'Probe not needed on localhost' error.

glusterd2 does this too. Right now, the error returned will be 'peer already in cluster' which is technically true.

> If peer match is found, just attach the new address to the matched peer, and send an update to the rest of the cluster. The matched peer isn't contacted at all.

I did not quite understand this.

> If a match isn't found, do a probe RPC (same as the ValidateAddPeer RPC here) to the address.
If the RPC returns an error, return the error to the CLI.

Already does this.

> Check the returned peer id with existing peers

Already does this as `isPeerInCluster` is called after `ValidateAddPeer`

> If a match exists, attach the new address to it
> If not, a new peer is created.

Multiple address per peer isn't handled here or elsewhere in the code. Requires a lot more effort.